pir/cpp/context.h

```diff
@@ -0,0 +1,114 @@
+//
+// Copyright 2020 the authors listed in CONTRIBUTORS.md
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+#ifndef PIR_CONTEXT_H_
+#define PIR_CONTEXT_H_
+
+#include "seal/seal.h"
+#include "util/statusor.h"
+
+namespace pir {
+
+using ::private_join_and_compute::StatusOr;
+
+class PIRContext {
+ public:
+  /**
+   * Creates a new context
+   * @returns TODO
+   **/
+  static StatusOr<std::unique_ptr<PIRContext>> Create();
+
+  /**
+   * Creates a new context from existing params
+   * @param[in] params Serialized PIR parameters
+   * @returns TODO
+   **/
+  static StatusOr<std::unique_ptr<PIRContext>> CreateFromParams(
+      const std::string& params);
+
+  /**
+   * Returns the serialized PIR parameters
+   * @returns TODO
+   **/
+  std::string SerializeParams() const;
+
+  /**
+   * Encodes a vector to a Plaintext
+   * @param[in] in Array to be encoded
+   * @returns TODO
+   **/
+  StatusOr<seal::Plaintext> Encode(const std::vector<uint64_t>& in);
+  /**
+   * Decodes a plaintext to a vector
+   * @param[in] in Plaintext to be decoded
+   * @returns TODO
+   **/
+  StatusOr<std::vector<uint64_t>> Decode(const seal::Plaintext& in);
+
+  /**
+   * Encodes, encrypts and serializes a vector
+   * @param[in] in Vector to be encrypted
+   * @returns TODO
+   **/
+  StatusOr<std::string> Encrypt(const std::vector<uint64_t>& in);
+
+  /**
+   * Deserializes, decrypts and decodes a vector
+   * @param[in] in Serialized ciphertext
+   * @returns TODO
+   **/
+  StatusOr<std::vector<uint64_t>> Decrypt(const std::string& in);
+
+  /**
+   * Serializes a ciphertext
+   * @param[in] in Ciphertext to be serialized
+   * @returns TODO
+   **/
+  StatusOr<std::string> Serialize(const seal::Ciphertext&);
+  /**
+   * Deserializes a ciphertext
+   * @param[in] in Serialized ciphertext
+   * @returns TODO
+   **/
+  StatusOr<seal::Ciphertext> Deserialize(const std::string& in);
+
+  /**
+   * Returns an Evaluator instance
+   * @returns TODO
+   **/
+  std::shared_ptr<seal::Evaluator>& Evaluator();
+
+ private:
+  PIRContext(const seal::EncryptionParameters&);
+
+  static seal::EncryptionParameters generateEncryptionParams(
+      uint32_t poly_modulus_degree = 4096, uint32_t plain_modulus = 1032193);
```

Comment:
Where does the plain modulus default come from?

Reply:
The plain modulus is a value that is valid for the BatchEncoder setup here. Citing from SEAL docs

> **BatchEncoder**
Provides functionality for CRT batching. If the polynomial modulus degree is N, and
the plaintext modulus is a prime number T such that T is congruent to 1 modulo 2N,
then BatchEncoder allows the plaintext elements to be viewed as 2-by-(N/2)
matrices of integers modulo T. Homomorphic operations performed on such encrypted
matrices are applied coefficient (slot) wise, enabling powerful SIMD functionality
for computations that are vectorizable. This functionality is often called "batching"
in the homomorphic encryption literature.
**Mathematical Background**
Mathematically speaking, if the polynomial modulus is X^N+1, N is a power of two, and
plain_modulus is a prime number T such that 2N divides T-1, then integers modulo T
contain a primitive 2N-th root of unity and the polynomial X^N+1 splits into n distinct
linear factors as X^N+1 = (X-a_1)*...*(X-a_N) mod T, where the constants a_1, ..., a_n
are all the distinct primitive 2N-th roots of unity in integers modulo T. The Chinese
Remainder Theorem (CRT) states that the plaintext space Z_T[X]/(X^N+1) in this case is
isomorphic (as an algebra) to the N-fold direct product of fields Z_T. The isomorphism
is easy to compute explicitly in both directions, which is what this class does.
Furthermore, the Galois group of the extension is (Z/2NZ)* ~= Z/2Z x Z/(N/2) whose
action on the primitive roots of unity is easy to describe. Since the batching slots
correspond 1-to-1 to the primitive roots of unity, applying Galois automorphisms on the
plaintext act by permuting the slots. By applying generators of the two cyclic
subgroups of the Galois group, we can effectively view the plaintext as a 2-by-(N/2)
matrix, and enable cyclic row rotations, and column rotations (row swaps).
**Valid Parameters**
Whether batching can be used depends on whether the plaintext modulus has been chosen
appropriately. Thus, to construct a BatchEncoder the user must provide an instance
of SEALContext such that its associated EncryptionParameterQualifiers object has the
flags parameters_set and enable_batching set to true.


We can make the generation more explicit, if it helps

Reply:
It is one of the reasons SEALPir fails with Seal 3.5

They are expecting batch mode enabled, and they provide 

```
int polymod = 2048; // N in the code
int polymod_log = 11; //logt in the code
plain_mod = 1 << (polymod_log + 1)+ 1;

```

but 2**12 + 1 is not prime.

Reply:
Perhaps we can use PlainModulus::Batching(poly_modulus_degree, 20) or something like that? Not sure what the bit size should be yet.